resource/cmake/CommonFunctions.cmake

```
@@ -0,0 +1,443 @@
+            ${_CONNEXTDDS_VERSION}
+        COMPONENTS
+            routing_service core
+        REQUIRED)
+
+    message(STATUS "definition added ${CONNEXTDDS_DEFINITIONS}")
+
+    # Configure Connext DDS libraries variables
+    set(ROUTING_SERVICE_CXX_LIBRARIES
+        RTIConnextDDS::routing_service_cpp2
+    )
+    set(ROUTING_SERVICE_C_LIBRARIES
+        RTIConnextDDS::routing_service_c
+    )
+endmacro()
+
+#[[
+
+m_add_subdirectory_if_exists
+--------------------------
+
+ * Type: macro
+ * Brief: adds a subdirectory if it exists
+ * Params:
+ ** _SUBDIRECTORY_DIR: path to the subdirectory to add
+ * How to use it:
+
+    m_add_subdirectory_if_exists("${CMAKE_CURRENT_SOURCE_DIR}/mySubdirectory")
+]]
+macro(m_add_subdirectory_if_exists _SUBDIRECTORY_DIR)
+    if (IS_DIRECTORY "${_SUBDIRECTORY_DIR}")
+        add_subdirectory("${_SUBDIRECTORY_DIR}")
+    endif()
+endmacro()
+
+#[[
+
+m_add_doc_if_exist
+----------------
+
+ * Type: macro
+ * Brief: adds the "doc" directory if it exists in the current source
+          directory. This folder is only added if "building doc" is enabled.
+ * Params: no parameters
+ * How to use it:
+
+    m_add_doc_if_exist()
+]]
+macro(m_add_doc_if_exist)
+    if (RTIGATEWAY_ENABLE_DOCS)
+        m_add_subdirectory_if_exists("${CMAKE_CURRENT_SOURCE_DIR}/doc")
+    endif()
+endmacro()
+
+#[[
+
+m_add_test_if_exist
+-----------------
+
+ * Type: macro
+ * Brief: adds the "test" directory if it exists in the current source
+          directory. This folder is only added if testing is enabled.
+ * Params: no parameters
+ * How to use it:
+
+    m_add_test_if_exist()
+]]
+macro(m_add_test_if_exist)
+    if (RTIGATEWAY_ENABLE_TESTS)
+        m_add_subdirectory_if_exists("${CMAKE_CURRENT_SOURCE_DIR}/test")
+    endif()
+endmacro()
+
+#[[
+
+add_plugin
+----------
+
+ * Type: function
+ * Brief: adds the corresponding plugin directories if they are added. These
+          directories are: the main plugin directory, related resources and
+          related examples for that plugin.
+ * Params:
+ ** ENABLE_PLUGIN: boolean that indicates that this plugin is enabled
+ ** PREFIX: plugin prefix
+ ** STAGING_DIR: staging directory name for the plugin
+ ** PLUGIN_DIR_NAME: name that identifies the folder of that plugin
+ ** PLUGIN_TYPE: indicates whether a plugin is an "ADAPTER", a "PROCESSOR" or a
+                 "TRANSFORMATION"
+ * Restrictions: CONNEXTDDS_DIR should be set
+ * How to use it:
+
+    add_plugin(
+        ENABLE_PLUGIN RTIGATEWAY_ENABLE_MODBUS
+        PREFIX "RTI_MODBUS"
+        STAGING_DIR "modbus"
+        PLUGIN_DIR_NAME "modbus"
+        PLUGIN_TYPE "ADAPTER"
+    )
+]]
+function(add_plugin)
+    set(BOOLEANS ENABLE_PLUGIN)
+    set(SINGLE_VALUE_ARGS PREFIX STAGING_DIR PLUGIN_DIR_NAME PLUGIN_TYPE)
+    set(MULTI_VALUE_ARGS)
+
+    cmake_parse_arguments(_COMMON_FUNCTIONS
+        "${BOOLEANS}"
+        "${SINGLE_VALUE_ARGS}"
+        "${MULTI_VALUE_ARGS}"
+        ${ARGN}
+    )
```

In general, I suggest using shorter names for both arguments and the local variables used to describe them (also use `_` to mark them as local):

```suggestion
function(rtigw_add_plugin)
    set(_BOOLEAN IF)
    set(_SINGLE_VALUE PREFIX STAGING DIR_NAME TYPE)
    set(_MULTI_VALUE)

    cmake_parse_arguments(_args
        "${_BOOLEAN}" "${_SINGLE_VALUE}" "${_MULTI_VALUE}" ${ARGN})
```

---

marked with `_`, I haven't renamed them because in the end, it doesn't really affect anything, (we only use them in the `cmake_parse_arguments`) function